### ISSUE TITLE
IList.traverse over disjunction is super slow

### ISSUE THREAD

**Issue by fommil** [scalaz/scalaz#1956]
This hacky way, which we could introduce as `def traverseDisjunction` or something on `IList` as a specialisation, is about x10 faster

```scala
object FastTraverse {

  implicit final class IListExtras[A](private val self: IList[A])
    extends AnyVal {
    def traverseDisjunction[E, B](f: A => E \/ B): E \/ IList[B] = {
      val res = self.map { a =>
        f(a) match {
          case -\/(err) => return -\/(err) // scalafix:ok
          case \/-(j) => j
        }
      }
      \/-(res)
    }
  }
}
```

**Comment by fommil** [comment]
looks like maybe a less hacky thing (that still isn't as fast) would be to implement an efficient `apply2` on disjunction.

**Comment by mcanlas** [comment]
Is this now solved given #1958 is merged?

**Comment by fommil** [comment]
yes it is :smile:
